File: player/windows/src/PetsGraph.Core/PetPackModels.cs
using System.Text.Json.Serialization;

namespace PetsGraph.Core;

public sealed record PetPackManifest
{
    public required string FormatVersion { get; init; }
    public required PackageIdentity Package { get; init; }
    public required PetIdentity Pet { get; init; }
    public required StageDefinition Stage { get; init; }
    public required CapabilityDeclaration Capabilities { get; init; }
    public required string Graph { get; init; }
    public required string Behavior { get; init; }
    public required string Integrity { get; init; }
}

public sealed record PackageIdentity
{
    public required string Id { get; init; }
    public required SemanticVersion ContentVersion { get; init; }
    public required string CreatedAt { get; init; }
}

public sealed record PetIdentity
{
    public required string Id { get; init; }
    public required string DisplayName { get; init; }
    public required string Species { get; init; }
}

public sealed record StageDefinition
{
    public required int[] ReferenceCanvasPx { get; init; }
    public required string Anchor { get; init; }
    public required double BaseDisplayHeight { get; init; }
    public required string DefaultNode { get; init; }
}

public sealed record CapabilityDeclaration
{
    public required string[] Required { get; init; }
    public required string[] Optional { get; init; }
}

public sealed record PetGraph
{
    public required string FormatVersion { get; init; }
    public required PetGraphNode[] Nodes { get; init; }
    public required PetGraphEdge[] Edges { get; init; }
}

public sealed record PetGraphNode
{
    public required string Id { get; init; }
    public required string Role { get; init; }
    public required string Scene { get; init; }
    public string? LoopClip { get; init; }
    public required bool AutonomousEligible { get; init; }
}

public sealed record PetGraphEdge
{
    public required string Id { get; init; }
    public required string From { get; init; }
    public required string To { get; init; }
    public required string Clip { get; init; }
    public required string InterruptPolicy { get; init; }
}

public sealed record PetBehavior
{
    public required string FormatVersion { get; init; }
    public required string Profile { get; init; }
    public required string DefaultNode { get; init; }
    public required BehaviorTiming Timing { get; init; }
    public required Dictionary<string, double> NodeWeights { get; init; }
    public required Dictionary<string, double> SceneWeights { get; init; }
}

public sealed record BehaviorTiming
{
    public required string Strategy { get; init; }
    public required Dictionary<string, double[]> DwellRangesSeconds { get; init; }
    public required bool AvoidImmediateRepeat { get; init; }
}

public sealed record PetClip
{
    public required string FormatVersion { get; init; }
    public required string Id { get; init; }
    public required string Type { get; init; }
    public required string EntryNode { get; init; }
    public required string ExitNode { get; init; }
    public required FrameRate FrameRate { get; init; }
    public required int FrameCount { get; init; }
    public required double DurationSeconds { get; init; }
    public required int[] SafeExitFrames { get; init; }
    public required ClipStage Stage { get; init; }
    public required ClipGeometry Geometry { get; init; }
    public required ClipPlayback Playback { get; init; }
    public required ClipProduction Production { get; init; }
    public required ClipRepresentation[] Representations { get; init; }
}

public sealed record FrameRate
{
    public required int Numerator { get; init; }
    public required int Denominator { get; init; }
    [JsonIgnore]
    public double FramesPerSecond => Numerator / (double)Denominator;
}

public sealed record ClipStage
{
    public required int[] ReferenceCanvasPx { get; init; }
    public required string Anchor { get; init; }
}

public sealed record ClipGeometry
{
    public required int[] CropPx { get; init; }
    public required int[] PresentationOffsetPx { get; init; }
}

public sealed record ClipPlayback
{
    public required bool NativeContinuousFrames { get; init; }
    public required double Rate { get; init; }
    public required string SpeedProcessing { get; init; }
}

public sealed record ClipProduction
{
    public required string RecipeDigest { get; init; }
    public required string ApprovalDigest { get; init; }
}

public sealed record ClipRepresentation
{
    public required string Id { get; init; }
    public required string Kind { get; init; }
    public required string Path { get; init; }
    public required string Encoding { get; init; }
    public required int WidthPx { get; init; }
    public required int HeightPx { get; init; }
    public required int BytesPerRow { get; init; }
    public required int FrameCount { get; init; }
    public required FrameRate FrameRate { get; init; }
    public required string Alpha { get; init; }
    public required string ColorSpace { get; init; }
    public required long Bytes { get; init; }
    public required string Sha256 { get; init; }
}

public sealed record PetPackIntegrity
{
    public required string FormatVersion { get; init; }
    public required string Algorithm { get; init; }
    public required IntegrityEntry[] Files { get; init; }
}

public sealed record IntegrityEntry
{
    public required string Path { get; init; }
    public required long Bytes { get; init; }
    public required string MediaType { get; init; }
    public required string Sha256 { get; init; }
}

public sealed record LoadedPetPack(
    string RuntimeRoot,
    PetPackManifest Manifest,
    PetGraph Graph,
    PetBehavior Behavior,
    IReadOnlyDictionary<string, PetClip> Clips,
    string ArchiveSha256,
    long ArchiveBytes)
{
    public string MediaPath(string clipId)
    {
        if (!Clips.TryGetValue(clipId, out var clip) || clip.Representations.Length != 1)
        {
            throw new PetPackException("missing_media", "clip has no baseline media");
        }
        return Path.Combine(RuntimeRoot, clip.Representations[0].Path.Replace('/', Path.DirectorySeparatorChar));
    }
}

public sealed record PetPackValidationReport(
    string PackageId,
    string PetId,
    string DisplayName,
    string Species,
    SemanticVersion ContentVersion,
    string ArchiveSha256,
    long ArchiveBytes,
    long UncompressedBytes,
    int EntryCount,
    int ClipCount,
    int NodeCount,
    int EdgeCount);

public sealed record ValidatedPetPack(LoadedPetPack Package, PetPackValidationReport Report);

File: player/macos/Sources/PetsGraphV1Core/PetPackModels.swift
import Foundation

public struct PetPackManifest: Codable, Equatable, Sendable {
  public let formatVersion: String
  public let package: PackageIdentity
  public let pet: PetIdentity
  public let stage: StageDefinition
  public let capabilities: CapabilityDeclaration
  public let graph: String
  public let behavior: String
  public let integrity: String
}

public struct PackageIdentity: Codable, Equatable, Sendable {
  public let id: String
  public let contentVersion: SemanticVersion
  public let createdAt: String
}

public struct PetIdentity: Codable, Equatable, Sendable {
  public let id: String
  public let displayName: String
  public let species: String
}

public struct StageDefinition: Codable, Equatable, Sendable {
  public let referenceCanvasPx: [Int]
  public let anchor: String
  public let baseDisplayHeight: Double
  public let defaultNode: String
}

public struct CapabilityDeclaration: Codable, Equatable, Sendable {
  public let required: [String]
  public let optional: [String]
}

public struct PetGraph: Codable, Equatable, Sendable {
  public let formatVersion: String
  public let nodes: [PetGraphNode]
  public let edges: [PetGraphEdge]
}

public struct PetGraphNode: Codable, Equatable, Sendable {
  public let id: String
  public let role: String
  public let scene: String
  public let loopClip: String?
  public let autonomousEligible: Bool
}

public struct PetGraphEdge: Codable, Equatable, Sendable {
  public let id: String
  public let from: String
  public let to: String
  public let clip: String
  public let interruptPolicy: String
}

public struct PetBehavior: Codable, Equatable, Sendable {
  public let formatVersion: String
  public let profile: String
  public let defaultNode: String
  public let timing: BehaviorTiming
  public let nodeWeights: [String: Double]
  public let sceneWeights: [String: Double]
}

public struct BehaviorTiming: Codable, Equatable, Sendable {
  public let strategy: String
  public let dwellRangesSeconds: [String: [Double]]
  public let avoidImmediateRepeat: Bool
}

public struct PetClip: Codable, Equatable, Sendable {
  public let formatVersion: String
  public let id: String
  public let type: String
  public let entryNode: String
  public let exitNode: String
  public let frameRate: FrameRate
  public let frameCount: Int
  public let durationSeconds: Double
  public let safeExitFrames: [Int]
  public let stage: ClipStage
  public let geometry: ClipGeometry
  public let playback: ClipPlayback
  public let production: ClipProduction
  public let representations: [ClipRepresentation]
}

public struct FrameRate: Codable, Equatable, Sendable {
  public let numerator: Int
  public let denominator: Int

  public var framesPerSecond: Double {
    Double(numerator) / Double(denominator)
  }
}

public struct ClipStage: Codable, Equatable, Sendable {
  public let referenceCanvasPx: [Int]
  public let anchor: String
}

public struct ClipGeometry: Codable, Equatable, Sendable {
  public let cropPx: [Int]
  public let presentationOffsetPx: [Int]
}

public struct ClipPlayback: Codable, Equatable, Sendable {
  public let nativeContinuousFrames: Bool
  public let rate: Double
  public let speedProcessing: String
}

public struct ClipProduction: Codable, Equatable, Sendable {
  public let recipeDigest: String
  public let approvalDigest: String
}

public struct ClipRepresentation: Codable, Equatable, Sendable {
  public let id: String
  public let kind: String
  public let path: String
  public let encoding: String
  public let widthPx: Int
  public let heightPx: Int
  public let bytesPerRow: Int
  public let frameCount: Int
  public let frameRate: FrameRate
  public let alpha: String
  public let colorSpace: String
  public let bytes: Int64
  public let sha256: String
}

public struct PetPackIntegrity: Codable, Equatable, Sendable {
  public let formatVersion: String
  public let algorithm: String
  public let files: [IntegrityEntry]
}

public struct IntegrityEntry: Codable, Equatable, Sendable {
  public let path: String
  public let bytes: Int64
  public let mediaType: String
  public let sha256: String
}

public struct LoadedPetPack: Sendable {
  public let runtimeRootURL: URL
  public let manifest: PetPackManifest
  public let graph: PetGraph
  public let behavior: PetBehavior
  public let clips: [String: PetClip]
  public let archiveSHA256: String

  public func mediaURL(for clipID: String) -> URL? {
    guard let path = clips[clipID]?.representations.first?.path else { return nil }
    return runtimeRootURL.appendingPathComponent(path, isDirectory: false)
  }
}

public struct PetPackValidationReport: Equatable, Sendable {
  public let packageID: String
  public let petID: String
  public let displayName: String
  public let species: String
  public let contentVersion: SemanticVersion
  public let archiveSHA256: String
  public let archiveBytes: Int64
  public let uncompressedBytes: Int64
  public let entryCount: Int
  public let clipCount: Int
  public let nodeCount: Int
  public let edgeCount: Int
}
